### DISCUSSION TITLE
Lots of open sockets

### DISCUSSION THREAD

**Discussion by cuu508** [emersion/maddy#596] [Q&A]
Hello,

I noticed that a server running maddy has lots of open sockets:

![image](https://github.com/foxcpp/maddy/assets/661859/43fbb50f-5fc3-417b-b71a-1b391a1a89c3)

(this is a screenshot from netdata, the number is *thousands* of sockets)

After restarting maddy the number drops close to zero and starts to slowly creep up again. Is this normal, and what might be causing this? Could there be potential problems if the number continues to rise?

Right now, with maddy relatively recently restarted if I run `lsof` and grep for `maddy`, I get:

```
lsof | grep maddy | wc -l
4837
```

Most of the entries in there look like so:

```
maddy     1450 1451 maddy                maddy   80u     sock                0,8      0t0      28175 protocol: TCP
maddy     1450 1451 maddy                maddy   81u     sock                0,8      0t0      28161 protocol: TCP
maddy     1450 1451 maddy                maddy   82u     sock                0,8      0t0      24235 protocol: TCP
maddy     1450 1451 maddy                maddy   83u     sock                0,8      0t0      32290 protocol: TCP
maddy     1450 1451 maddy                maddy   84u     sock                0,8      0t0      24215 protocol: TCP
maddy     1450 1451 maddy                maddy   85u     sock                0,8      0t0      32287 protocol: TCP
maddy     1450 1451 maddy                maddy   86u     sock                0,8      0t0      43947 protocol: TCP
```

Is this maddy keeping a cache of idle connections or something else?

Not sure if related, but looking at `internal/smtpconn/pool/pool.go`, it looks like cache entries are evicted from cache only when the cache is already full:

```go
		// Garbage-collect stale buckets.
		if len(p.keys) == p.cfg.MaxKeys {
			for k, v := range p.keys {
				if v.lastUse+p.cfg.StaleKeyLifetimeSec > time.Now().Unix() {
					continue
				}

				close(v.c)
				for conn := range v.c {
					conn.Close()
				}
				delete(p.keys, k)
			}
		}
```

MaxKeys is 20000 and the default MaxConnsPerKey is 10, so the cache could contain up to 200K connections, isn't this too much?

**Comment by foxcpp** [comment]
Addressed in b8ff116. Stale connections are closed every minute now, MaxKeys is reduced to 5000, MaxConnsPerKey to 5.

**Reply by cuu508** [reply]
Awesome, thank you for looking into this!

**Reply by cuu508** [reply]
I've been running a maddy build with this patch included for a while now, and can confirm that the number of open sockets stays low and does not climb any more.
